GPU/GPUTracking/DataTypes/GPUTRDTrack.h

```diff
@@ -71,57 +78,55 @@ class GPUTRDTrack_t : public T
   GPUd() GPUTRDTrack_t(const T& t);
   GPUd() GPUTRDTrack_t& operator=(const GPUTRDTrack_t& t);
 
-  GPUd() int GetNlayers() const;
-  GPUd() int GetTracklet(int iLayer) const;
-  GPUd() int GetTPCtrackId() const { return mTPCTrackId; }
-  GPUd() int GetNtracklets() const { return mNTracklets; }
-  GPUd() int GetNtrackletsOffline(int type) const { return mNTrackletsOffline[type]; }
-  GPUd() int GetLabelOffline() const { return mLabelOffline; }
-  GPUd() int GetLabel() const { return mLabel; }
-  GPUd() float GetChi2() const { return mChi2; }
-  GPUd() float GetReducedChi2() const { return GetNlayers() == 0 ? mChi2 : mChi2 / GetNlayers(); }
-  GPUd() float GetMass() const { return mMass; }
-  GPUd() bool GetIsStopped() const { return mIsStopped; }
-  GPUd() bool GetIsFindable(int iLayer) const { return mIsFindable[iLayer]; }
-  GPUd() int GetTrackletIndex(int iLayer) const { return GetTracklet(iLayer); }
-  GPUd() int GetNmissingConsecLayers(int iLayer) const;
-
-  GPUd() void AddTracklet(int iLayer, int idx)
+  // attach a tracklet to this track; this overwrites the mIsFindable flag to true for this layer
+  GPUd() void addTracklet(int iLayer, int idx) { mAttachedTracklets[iLayer] = idx; }
+
+  // getters
+  GPUd() int getNlayersFindable() const;
+  GPUd() int getTracklet(int iLayer) const { return mAttachedTracklets[iLayer]; }
+#ifdef HAVE_O2HEADERS
```

Comment:
This is not so save, since the HAVE_O2HEADERS is not in the PUBLIC definitions.
I would put this function inside of `TrackTRD.h`.

And I would rename it into getReferenceTrackGlobalId(), since it is not the global Id of this track, but of the TPC or TPC-ITS reference. You can also put a comment `// return GlobalID of reference track in its original container`.

Reply:
OK, thinking about that, it is not so simple, since `TrackTRD.h` contains only an alias using the `using` keyword.
What you can do is: You place only the declaration of the function here in GPUTRDTrack.h. You might need to forward-declare GlobalTrackId. Then you can put the function body as inline function inside TrackTRD.h. This way you make sure it is always available for O2, but not otherwise.

Reply:
But if I do that then how can I define the function for `o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTPCGMTrackParam> >`?

And also it does not work for the O2 track type for some reason:

`/home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/TRD/include/DataFormatsTRD/TrackTRD.h:30:48: error: definition of 'GTrackID o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >::getReferenceTrackGlobalId() const' is not in namespace enclosing 'o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >' [-fpermissive]
   30 | GTrackID TrackTRD::getReferenceTrackGlobalId() const
      |                                                ^~~~~
/home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/TRD/include/DataFormatsTRD/TrackTRD.h:30:10: error: no declaration matches 'GTrackID o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >::getReferenceTrackGlobalId() const'
   30 | GTrackID TrackTRD::getReferenceTrackGlobalId() const
      |          ^~~~~~~~
In file included from /home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/TRD/include/DataFormatsTRD/TrackTRD.h:17,
                 from /home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h:27,
                 from /home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx:17:
/home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/GPU/GPUTracking/DataTypes/GPUTRDTrack.h:91:41: note: candidate is: 'o2::dataformats::GlobalTrackID o2::gpu::GPUTRDTrack_t<T>::getReferenceTrackGlobalId() const [with T = o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack>]'
   91 |   GPUd() o2::dataformats::GlobalTrackID getReferenceTrackGlobalId() const;
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/GPU/GPUTracking/DataTypes/GPUTRDTrack.h:59:7: note: 'class o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >' defined here
   59 | class GPUTRDTrack_t : public T
      |       ^~~~~~~~~~~~~`

And when I change the definition inside `TrackTRD` to `o2::dataformats::GlobalTrackID o2::gpu::GPUTRDTrack_t<T>::getReferenceTrackGlobalId() const` then I still get a compiler error:
`/home/oschmidt/alice/sw/SOURCES/O2/gputrdtracking/0/DataFormats/Detectors/TRD/include/DataFormatsTRD/TrackTRD.h:30:65: error: definition of 'GTrackID o2::gpu::GPUTRDTrack_t<T>::getReferenceTrackGlobalId() const' is not in namespace enclosing 'o2::gpu::GPUTRDTrack_t<T>' [-fpermissive]
   30 | GTrackID o2::gpu::GPUTRDTrack_t<T>::getReferenceTrackGlobalId() const
      |                                                                 ^~~~~
ninja: build stopped: subcommand failed.`

My declaration inside `GPUTRDTrack.h` looks as follows:
`GPUd() o2::dataformats::GlobalTrackID getReferenceTrackGlobalId() const;`
Is the problem the `GPUd()` macro?